Numpy module imported twice #4218

Amrithasuresh opened the pull request:
Hi,

I committed changes for this issue https://github.com/Theano/Theano/issues/4218.

This is my first time to github contribution.

As per your suggestion in this page https://github.com/Theano/Theano/issues/4218

I did following things

1) converted "import numpy as N" to "import numpy as np"
2) change "import numpy" to "import numpy as np"
3) Made one commit per file.
4) I did manually. I have not used sed or anything automated.
5) I did not touch any c code.

Hope this helps. Thank you.

affanv14 commented on theano/gradient.py:
```diff
@@ -6,7 +6,7 @@
 import warnings
 
-import numpy  # for numeric_grad
+import numpy as np # for numeric_grad
```
Since you changed it to np,you need to remove line 22.

Amrithasuresh replied:
>"you need to change numpy to np in elemwise.py line 490"

Thank you for your patience. I have changed numpy to np in elemwise.py in lines 490 and 507. 

Do I need to check all other remaining files?

nouiz commented:
Sorry for not giving comments earlier, but we where pretty busy with the release. Now we start to merge again in the master of Theano changes that we won't include in Theano 0.9.

This PR need a rebase, can you do it? If you need help, tell us.

thanks

Amrithasuresh commented:
@nouiz . Thank you for the reply.

> This PR need a rebase, can you do it? If you need help, tell us.

I can do it. I need help. Can you please explain a bit? Why we need a rebase here? 

I understand from a [stackoverflow comment](http://stackoverflow.com/questions/2715085/rebasing-and-what-does-one-mean-by-rebasing-pushed-commits)  that a rebase is

"When you rebase stuff, you’re abandoning existing commits and creating new ones that are similar but different.".

Thank you.

Amrithasuresh commented:
@nouiz 

I have a created PR  https://github.com/Theano/Theano/pull/5714. Now, there are no conflicts. Please have a look. Thank you.


**Steps I have followed**

1) git clone https://github.com/Amrithasuresh/Theano.git
2) cd Theano/
3) git checkout -b my-branch
4) git remote add Theano https://github.com/Theano/Theano.git
5) git fetch Theano
6) git merge-base my-branch Theano/master
7) git rebase -i 19540d
8) Modified conflicting flies
9) Added the corrected files (git add file1,file2 etc..)
10) git rebase --continue


I have followed this [tutorial](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request)

nouiz commented:
All the commits have been merged into one commit. This isn't great. From
the instruction you have written, I don't see the problem. Did you do git
rebase only once or multiple time? Mostly, did you do a loop over step 8, 9
and 10?

On Wed, Mar 15, 2017 at 5:01 AM Suresh Panneerselvam <
notifications@github.com> wrote:

> @nouiz <https://github.com/nouiz>
>
> I have a created PR #5714 <https://github.com/Theano/Theano/pull/5714>.
> Now, there are no conflicts. Please have a look. Thank you.
>
> *Steps I have followed*
>
>    1. git clone https://github.com/Amrithasuresh/Theano.git
>    2. cd Theano/
>    3. git checkout -b my-branch
>    4. git remote add Theano https://github.com/Theano/Theano.git
>    5. git fetch Theano
>    6. git merge-base my-branch Theano/master
>    7. git rebase -i 19540d
>    8. Modified conflicting flies
>    9. Added the corrected files (git add file1,file2 etc..)
>    10. git rebase --continue
>
> I have followed this tutorial
> <https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5640#issuecomment-286680026>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-4PGSg6-flNcNwfGp9Y6PyF65E0kks5rl6jcgaJpZM4MPr0I>
> .
>

Amrithasuresh commented:
>All the commits have been merged into one commit. This isn't great.

As I mentioned in the above steps, I gave the command [ 7)  git rebase -I 19540d]. A nano text editor opened. I squashed all the commits except first one. I misunderstood that one of the purposes of rebasing a pull request is to squash the similar small commits into one. What shall I do now? Shall I rebase the pull request again without squash? Will it work? Thank you for your patience.

nouiz commented:
Do a new rebase from your original PR without squashing

rebasing can be used for many reason. This include squashing, but for this
type of PR, we prefer to have many commits. It make a big difference for
the review.

Why you need to rebase is that there is conflict. rebasing is a way to fix
conflict (as merge, but we don't want merge inside PR).

thanks

On Wed, Mar 15, 2017 at 10:44 PM Suresh Panneerselvam <
notifications@github.com> wrote:

> All the commits have been merged into one commit. This isn't great.
>
> As I mentioned in the above steps, I gave the command [ 7) git rebase -I
> 19540d]. A nano text editor opened. I squashed all the commits except first
> one. I misunderstood that one of the purposes of rebasing a pull request is
> to squash the similar small commits into one. What shall I do now? Shall I
> rebase the pull request again without squash? Will it work? Thank you for
> your patience.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5640#issuecomment-286941908>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-zLFFC8pVz2ZHA6aXKl6PLJvDncyks5rmKH_gaJpZM4MPr0I>
> .
>

Amrithasuresh commented:
@nouiz It was hard for me to fix all the files together. So, the commit messages cannot be clear and it is messy. I guess it will be a headache for reviewing this [PR](https://github.com/Theano/Theano/pull/5714). But, now I know how and where to do changes. In the weekend I will do a fresh fork and do fix the files one by one. That will be clear and easy. Thank you.

nouiz commented:
Just to be sure of where you go about this. Can you make 1 PR with the first 4-5 commits with fixed conflict. Then when it get merged, you can go on with the next 4-5 commits?

This way, I think each small PR will conflict less often and it will help get this merged.

thanks.

Amrithasuresh commented:
>Can you make 1 PR with the first 4-5 commits with fixed conflict. Then when it get merged, you can go on with the next 4-5 commits?

@nouiz I will do the same way. I will delete this PR and the other [PR](https://github.com/Theano/Theano/pull/5714). They are messy and unclear. I will do a fresh fork and commit 5 files per PR. Thank you.

Amrithasuresh commented:
As per suggestions, I did the changes for few files https://github.com/Theano/Theano/pull/5752
